### PR TITLE
gh-113688: Split up gcmodule.c

### PR DESCRIPTION
**Pull request by colesbury**
This splits part of Modules/gcmodule.c of into Python/gc.c, which now contains the core garbage collection implementation. The Python module remain in the Modules/gcmodule.c file.

* Issue: gh-113688

### CODE REVIEW

**Comment by colesbury** [comment]
Note for reviewers: The PR is split in two commits in an attempt to make it easier to review. The first commit simply moves `Modules/gcmodule.c` to `Python/gc.c` without changes. The second commit pulls out the Python package back into `Modules/gcmodule.c` and fixes things up.

I recommend reviewing the second commit: https://github.com/python/cpython/pull/113715/commits/fa068b5b9ca51e75d3bedf7331b8aca43b6638b1 "Split up gcmodule.c".

Here's the summary of changes beyond simply moving functions from one file to another:
* Move logic for freezing, unfreezing, getting all GC objects, etc. to `Python/gc.c`. See changes to [`pycore_gc.h`](https://github.com/python/cpython/pull/113715/files#diff-0e64649e0124da22bbdd808a1902470f99752765073e6f1eb8bbd734ef1f4255). This avoids requiring many of the GC list operations in `Modules/gcmodule.c`. 
* Move the `DEBUG_STATS`, etc. constants to `pycore_gc.h` and prefix them as e.g., `_PyGC_DEBUG_STATS`. The Python constants are not renamed (i.e., `gc.DEBUG_STATS`). They are used in both `Modules/gcmodule.c` and `Python/gc.c`.

**Comment by nascheme** [comment]
Looks okay to me.  I'd suggest it's time to revise that comment at the top of the file.  It's not so useful anymore and someone can use git log to see contributors and the history.  Maybe add a link to `https://devguide.python.org/internals/garbage-collector/` instead of links to those mailing list posts.  E.g. you could put this at the top of the gc.c file instead:

```
//  This implements the reference cycle garbage collector.
//  The Python module inteface to the collector is in gcmodule.c.
//  See https://devguide.python.org/internals/garbage-collector/
```

File: Python/gc.c [hunk 1]
```diff
@@ -0,0 +1,1939 @@
...
+             * set but can be deallocated as a result of breaking the
+             * reference cycle.  If we don't clear the weakref, the callback
+             * will run and potentially cause a crash.  See bpo-38006 for
+             * one example.
+             */
+            _PyWeakref_ClearRef((PyWeakReference *)op);
+        }
+
+        if (! _PyType_SUPPORTS_WEAKREFS(Py_TYPE(op)))
+            continue;
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
        if (! _PyType_SUPPORTS_WEAKREFS(Py_TYPE(op))) {
            continue;
        }
```

File: Python/gc.c [hunk 2]
```diff
@@ -0,0 +1,1939 @@
...
+        wr = (PyWeakReference *)op;
+        callback = wr->wr_callback;
+        _PyObject_ASSERT(op, callback != NULL);
+
+        /* copy-paste of weakrefobject.c's handle_callback() */
+        temp = PyObject_CallOneArg(callback, (PyObject *)wr);
+        if (temp == NULL)
+            PyErr_WriteUnraisable(callback);
+        else
+            Py_DECREF(temp);
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
        if (temp == NULL) {
            PyErr_WriteUnraisable(callback);
        }
        else {
            Py_DECREF(temp);
        }
```

File: Python/gc.c [hunk 3]
```diff
@@ -0,0 +1,1939 @@
...
+     * care.
+     */
+    gc_list_init(&seen);
+
+    while (!gc_list_is_empty(collectable)) {
+        PyGC_Head *gc = GC_NEXT(collectable);
+        PyObject *op = FROM_GC(gc);
+        gc_list_move(gc, &seen);
+        if (!_PyGCHead_FINALIZED(gc) &&
+                (finalize = Py_TYPE(op)->tp_finalize) != NULL) {
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
        if (!_PyGCHead_FINALIZED(gc) &&
            (finalize = Py_TYPE(op)->tp_finalize) != NULL)
        {
```

File: Python/gc.c [hunk 4]
```diff
@@ -0,0 +1,1939 @@
...
+               thusly: "each full garbage collection is more and more costly as the
+               number of objects grows, but we do fewer and fewer of them").
+
+               This heuristic was suggested by Martin von Löwis on python-dev in
+               June 2008. His original analysis and proposal can be found at:
+               http://mail.python.org/pipermail/python-dev/2008-June/080579.html
+            */
+            if (i == NUM_GENERATIONS - 1
+                && gcstate->long_lived_pending < gcstate->long_lived_total / 4)
+                continue;
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
            if (i == NUM_GENERATIONS - 1
                && gcstate->long_lived_pending < gcstate->long_lived_total / 4)
            {
                continue;
            }
```

File: Python/gc.c [hunk 5]
```diff
@@ -0,0 +1,1939 @@
...
+    }
+
+    if (PyDTrace_GC_START_ENABLED())
+        PyDTrace_GC_START(generation);
+
+    /* update collection and allocation counters */
+    if (generation+1 < NUM_GENERATIONS)
+        gcstate->generations[generation+1].count += 1;
+    for (i = 0; i <= generation; i++)
+        gcstate->generations[i].count = 0;
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
    if (PyDTrace_GC_START_ENABLED()) {
        PyDTrace_GC_START(generation);
    }

    /* update collection and allocation counters */
    if (generation+1 < NUM_GENERATIONS) {
        gcstate->generations[generation+1].count += 1;
    }
    for (i = 0; i <= generation; i++) {
        gcstate->generations[i].count = 0;
    }
```

File: Python/gc.c [hunk 6]
```diff
@@ -0,0 +1,1939 @@
...
+    for (i = 0; i < generation; i++) {
+        gc_list_merge(GEN_HEAD(gcstate, i), GEN_HEAD(gcstate, generation));
+    }
+
+    /* handy references */
+    young = GEN_HEAD(gcstate, generation);
+    if (generation < NUM_GENERATIONS-1)
+        old = GEN_HEAD(gcstate, generation+1);
+    else
+        old = young;
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
    if (generation < NUM_GENERATIONS-1) {
        old = GEN_HEAD(gcstate, generation+1);
    }
    else {
        old = young;
    }
```

File: Python/gc.c [hunk 7]
```diff
@@ -0,0 +1,1939 @@
...
+}
+
+static int
+referrersvisit(PyObject* obj, void *arg)
+{
+    PyObject *objs = arg;
+    Py_ssize_t i;
+    for (i = 0; i < PyTuple_GET_SIZE(objs); i++)
+        if (PyTuple_GET_ITEM(objs, i) == obj)
+            return 1;
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
    for (i = 0; i < PyTuple_GET_SIZE(objs); i++) {
        if (PyTuple_GET_ITEM(objs, i) == obj) {
            return 1;
        }
    }
```

File: Python/gc.c [hunk 8]
```diff
@@ -0,0 +1,1939 @@
...
+    traverseproc traverse;
+    for (gc = GC_NEXT(list); gc != list; gc = GC_NEXT(gc)) {
+        obj = FROM_GC(gc);
+        traverse = Py_TYPE(obj)->tp_traverse;
+        if (obj == objs || obj == resultlist)
+            continue;
+        if (traverse(obj, referrersvisit, objs)) {
+            if (PyList_Append(resultlist, obj) < 0)
+                return 0; /* error */
+        }
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
        if (obj == objs || obj == resultlist) {
            continue;
        }
        if (traverse(obj, referrersvisit, objs)) {
            if (PyList_Append(resultlist, obj) < 0) {
                return 0; /* error */
            }
        }
```

File: Python/gc.c [hunk 9]
```diff
@@ -0,0 +1,1939 @@
...
+    GCState *gcstate = &interp->gc;
+    if (!(gcstate->debug & _PyGC_DEBUG_SAVEALL)
+        && gcstate->garbage != NULL && PyList_GET_SIZE(gcstate->garbage) > 0) {
+        const char *message;
+        if (gcstate->debug & _PyGC_DEBUG_UNCOLLECTABLE)
+            message = "gc: %zd uncollectable objects at " \
+                "shutdown";
+        else
+            message = "gc: %zd uncollectable objects at " \
+                "shutdown; use gc.set_debug(gc.DEBUG_UNCOLLECTABLE) to list them";
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
        if (gcstate->debug & _PyGC_DEBUG_UNCOLLECTABLE) {
            message = "gc: %zd uncollectable objects at shutdown";
        }
        else {
            message = "gc: %zd uncollectable objects at shutdown; " \
                "use gc.set_debug(gc.DEBUG_UNCOLLECTABLE) to list them";
        }
```

File: Python/gc.c [hunk 10]
```diff
@@ -0,0 +1,1939 @@
...
+        /* PyErr_WarnFormat does too many things and we are at shutdown,
+           the warnings module's dependencies (e.g. linecache) may be gone
+           already. */
+        if (PyErr_WarnExplicitFormat(PyExc_ResourceWarning, "gc", 0,
+                                     "gc", NULL, message,
+                                     PyList_GET_SIZE(gcstate->garbage)))
+            PyErr_WriteUnraisable(NULL);
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
        if (PyErr_WarnExplicitFormat(PyExc_ResourceWarning, "gc", 0,
                                     "gc", NULL, message,
                                     PyList_GET_SIZE(gcstate->garbage)))
        {
            PyErr_WriteUnraisable(NULL);
        }
```

File: Python/gc.c [hunk 11]
```diff
@@ -0,0 +1,1939 @@
...
+        if (gcstate->debug & _PyGC_DEBUG_UNCOLLECTABLE) {
+            PyObject *repr = NULL, *bytes = NULL;
+            repr = PyObject_Repr(gcstate->garbage);
+            if (!repr || !(bytes = PyUnicode_EncodeFSDefault(repr)))
+                PyErr_WriteUnraisable(gcstate->garbage);
```
Comment (erlend-aasland) [inline Python/gc.c]:
```suggestion
            if (!repr || !(bytes = PyUnicode_EncodeFSDefault(repr))) {
                PyErr_WriteUnraisable(gcstate->garbage);
            }
```

**Review by erlend-aasland — APPROVED** [review]
Thanks, this is a nice refactor; the resulting extension module is a lot easier to read now.

Nitpick: since we're touching these lines, we now have a golden opportunity to align the new Python/gc.c file with PEP-7, so I suggest we do so.

**Comment by pablogsal** [comment]
> Looks okay to me. I'd suggest it's time to revise that comment at the top of the file. It's not so useful anymore and someone can use git log to see contributors and the history. Maybe add a link to `https://devguide.python.org/internals/garbage-collector/` instead of links to those mailing list posts. E.g. you could put this at the top of the gc.c file instead:
> 
> ```
> //  This implements the reference cycle garbage collector.
> //  The Python module inteface to the collector is in gcmodule.c.
> //  See https://devguide.python.org/internals/garbage-collector/
> ```

Related to this although not a blocker for this PR: let's ensure we document the noggin parts also in  https://devguide.python.org/internals/garbage-collector/

**Review by pablogsal — APPROVED** [review]
I made a first round of review and everything looks good to me.

I would advise to keep this PR at minimum changes (so let's avoid the temptation to do refactors 😉  ) to keep the split as "pure" as possible.

**Comment by colesbury** [comment]
@pablogsal - that makes sense. Should I revert the style changes and put them out as a subsequent PR?

**Comment by pablogsal** [comment]
Nah, don't worry, I agree it was a good opportunity but I was being careful that we end putting a lot more of other stylistic changes or similar

**Comment by pablogsal** [comment]
False positive, the builder is out of disk space

**Pull request merged**